File: scripts/node/parsers/permsetParser.js
/**
 * permsetParser.js
 * Reads Salesforce Permission Set metadata from SFDX source format.
 */

'use strict';

const fs   = require('fs');
const path = require('path');

function extractTag(xml, tag) {
  const m = xml.match(new RegExp(`<${tag}[^>]*>([\\s\\S]*?)<\\/${tag}>`));
  return m ? m[1].trim() : null;
}

function countTags(xml, tag) {
  const re = new RegExp(`<${tag}[^>]*>`, 'g');
  return (xml.match(re) || []).length;
}

/**
 * @param {string} permsetDir  - path to the permissionsets/ folder
 * @returns {{ permsets: Array<{ apiName, label, description, objectPerms, fieldPerms }> }}
 */
async function parsePermsets(permsetDir) {
  const permsets = [];

  if (!fs.existsSync(permsetDir)) return { permsets };

  const files = fs.readdirSync(permsetDir).filter(f => f.endsWith('.permissionset-meta.xml'));

  for (const file of files) {
    const xml     = fs.readFileSync(path.join(permsetDir, file), 'utf8');
    const apiName = file.replace('.permissionset-meta.xml', '');

    permsets.push({
      apiName,
      label:       extractTag(xml, 'label'),
      description: extractTag(xml, 'description'),
      objectPerms: countTags(xml, 'objectPermissions'),
      fieldPerms:  countTags(xml, 'fieldPermissions'),
    });
  }

  return { permsets };
}

module.exports = { parsePermsets };

File: scripts/python/parsers/permset_parser.py
"""
permset_parser.py
Reads Salesforce Permission Set metadata from SFDX source format.
Uses stdlib only (re, pathlib).
"""

import re
from pathlib import Path


def _extract_tag(xml: str, tag: str):
    m = re.search(rf'<{tag}[^>]*>([\s\S]*?)</{tag}>', xml)
    return m.group(1).strip() if m else None


def _count_tags(xml: str, tag: str) -> int:
    return len(re.findall(rf'<{tag}[^>]*>', xml))


def parse_permsets(permset_dir: str) -> dict:
    """
    Parse all permission sets under permset_dir.
    Returns {'permsets': [{'apiName', 'label', 'description', 'objectPerms', 'fieldPerms'}]}
    """
    permsets = []
    p = Path(permset_dir)

    if not p.exists():
        return {'permsets': permsets}

    for ff in sorted(p.glob('*.permissionset-meta.xml')):
        xml = ff.read_text(encoding='utf-8')
        api_name = ff.name.replace('.permissionset-meta.xml', '')
        permsets.append({
            'apiName': api_name,
            'label': _extract_tag(xml, 'label'),
            'description': _extract_tag(xml, 'description'),
            'objectPerms': _count_tags(xml, 'objectPermissions'),
            'fieldPerms': _count_tags(xml, 'fieldPermissions'),
        })

    return {'permsets': permsets}
